utils/stall_free.hh

```diff
@@ -57,14 +59,176 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
+// The containers can be re-used after clear_gently
+// or may be destroyed.  But unlike e.g. std::vector::clear(),
+// clear_gently will not necessarily keep the object allocation.
+
+template <typename T>
+concept HasClearGentlyMethod = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept SharedPointer = SmartPointer<T> && requires (T x) {
+    { x.use_count() } -> std::convertible_to<long>;
+};
+
+template <typename T>
+concept StringLike = requires (T x) {
+    std::is_same_v<typename T::traits_type, std::char_traits<typename T::value_type>>;
+};
+
+template <typename T>
+concept Iterable = requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.begin() } -> std::same_as<typename T::iterator>;
+    { x.end() } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept Sequence = Iterable<T> && requires (T x, size_t n) {
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept TriviallyClearableSequence =
+    Sequence<T> && std::is_trivially_destructible_v<typename T::value_type> && !HasClearGentlyMethod<typename T::value_type>;
+
+template <typename T>
+concept Container = Iterable<T> && requires (T x, typename T::iterator it) {
+    { x.erase(it) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept MapLike = Container<T> && requires (T x) {
+    std::is_same_v<typename T::value_type, std::pair<typename T::key_type, typename T::mapped_type>>;
+};
+
+template <HasClearGentlyMethod T>
+future<> clear_gently(T& o) noexcept;
+
+template <SharedPointer T>
+future<> clear_gently(T& o) noexcept;
+
+template <SmartPointer T>
+requires (!SharedPointer<T>)
+future<> clear_gently(T& o) noexcept;
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <typename T>
+requires (StringLike<T> || TriviallyClearableSequence<T>)
+future<> clear_gently(T& s) noexcept;
+
+template <Sequence T>
+requires (!StringLike<T> && !TriviallyClearableSequence<T>)
+future<> clear_gently(T& v) noexcept;
+
+template <MapLike T>
+future<> clear_gently(T& c) noexcept;
+
+template <Container T>
+requires (!StringLike<T> && !Sequence<T> && !MapLike<T>)
+future<> clear_gently(T& c) noexcept;
+
+namespace internal {
+
+template <typename T>
+concept HasClearGentlyImpl = requires (T x) {
+    { clear_gently(x) } -> std::same_as<future<>>;
+};
+
+template <typename T>
+requires HasClearGentlyImpl<T>
+future<> clear_gently(T& x) noexcept {
+    return utils::clear_gently(x);
+}
+
+// This default implementation of clear_gently
+// is required to "terminate" recursive clear_gently calls
+// at trivial objects
+template <typename T>
+requires (!HasClearGentlyImpl<T>)
```

Comment:
No need for the negative constraint, the compiler prefers more constrained options.

Reply:
Yup.  I got paranoid.

Reply:
Fixed in v7